src/main/java/cbot/io/Parser.java

```diff
@@ -0,0 +1,362 @@
+     * @see Command
+     */
+    public Parser(String input) throws PoorInputException {
+        this.c = null;
+        this.text = "";
+
+        boolean matchFound = false;
+
+        if (input.contains(Task.SEP)) {
+            throw new BadInputException("Please avoid using: \"" + Task.SEP + "\"");
+        }
+
+        for (Command c : Command.values()) {
+            if (c.matches(input)) {
+                this.c = c;
+                this.text = c.getText(input);
+                matchFound = true;
+                break;
+            }
+        }
+
+        if (!matchFound) {
+            throw new PoorInputException("Sorry, I don't recognize that command :<");
+        }
+    }
+
+    /**
+     * Returns true if the stored Command is the BYE Command.
+     *
+     * @return true if the stored Command is BYE.
+     */
+    public boolean isBye() {
+        return (this.c == Command.BYE);
+    }
+
+    /**
+     * Returns true if the stored Command requires the task list to be saved.
+     *
+     * @return Whether the task list needs to be saved after the Command.
+     */
+    public boolean needSave() {
+        return this.c.needSave();
+    }
+
+    /**
+     * Processes the Command and the rest of the input. If needed, the TaskList is modified and the UI is given
+     * outputs to print. The main bulk of Cbot's parsing is done here.
+     *
+     * @param tl The current list of tasks.
+     * @return The String response to the command.
+     * @throws PoorInputException If the input text is improper or erroneous.
+     * @throws DateTimeParseException If some provided datetime is not in a recognized format.
+     */
+    public String respond(TaskList tl) throws PoorInputException, DateTimeParseException {
+        String output;
+
+        if (this.c.isMissingText(this.text)) {
+            throw new PoorInputException("That command needs an input");
+        }
+
+        switch (this.c) {
+        case BYE:
+            output = UI.sayBye();
+            break;
+
+        case LIST:
+            if (tl.getCount() == 0) {
+                output = UI.say("Freedom! You have no tasks :D");
+            } else {
+                output = UI.say("Here's what you have:\n")
+                        + UI.printMany(tl.listTasks());
+            }
+            break;
+
+        case MARK:
+            try {
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
+                }
+
+                output = UI.say(tl.mark(num));
+
+            } catch (NumberFormatException e) {
+                throw new BadInputException("Invalid index!");
+            }
+            break;
+
+        case UNMARK:
+            try {
+                int num = Integer.parseInt(this.text);
+
+                if (tl.notInRange(num)) {
+                    throw new PoorInputException(tl.rangeError(num));
```

Comment:
This part seems very heavily indented. Perhaps the overall structure can be reorganised?

Reply:
Hi! Thanks so much for the comments.
Could I ask what you're referring to here? Just this `case UNMARK:` or the whole method? I'm not too sure how I can reorganize this `try`-`catch` block :(